composer/algorithms/factorize/factorize.py

```diff
@@ -0,0 +1,160 @@
+
+    def _maybe_replace_conv2d(module: torch.nn.Conv2d,
+                              module_index: int,
+                              min_channels: int = min_channels,
+                              latent_channels: FractionOrInt = latent_channels) -> Optional[torch.nn.Module]:
+        wide_enough = min(module.out_channels, module.in_channels) >= min_channels
+        if factorizing_could_speedup(module, latent_channels) and wide_enough:
+            return FactorizedConv2d.from_conv2d(module, module_index, latent_channels=latent_channels)
+        return None  # not enough rank reduction to be worth it
+
+    ret = _replace_module_class_in_model(model, torch.nn.Conv2d, _maybe_replace_conv2d)
+    _python_log_surgery_result(model, FactorizedConv2d)
+    return ret
+
+
+def factorize_linear_modules(model: torch.nn.Module, min_features: int, latent_features: FractionOrInt):
+
+    def _maybe_replace_linear(module: torch.nn.Linear,
+                              module_index: int,
+                              min_features: int = min_features,
+                              latent_features: FractionOrInt = latent_features) -> Optional[torch.nn.Module]:
+        wide_enough = min(module.in_features, module.out_features) >= min_features
+        if factorizing_could_speedup(module, latent_features) and wide_enough:
+            return FactorizedLinear.from_linear(module, module_index, latent_features=latent_features)
+        return None  # not enough rank reduction to be worth it
+
+    ret = _replace_module_class_in_model(model, torch.nn.Linear, _maybe_replace_linear)
+    _python_log_surgery_result(model, FactorizedLinear)
+    return ret
+
+
+@dataclass
+class FactorizeHparams(AlgorithmHparams):
+    factorize_convs: bool = hp.optional(
+        doc='Whether to factorize convolutional layers',
+        default=_DEFAULT_SHOULD_FACTORIZE_CONVS,
+    )
+    factorize_linears: bool = hp.optional(
+        doc='Whether to factorize linear layers',
+        default=_DEFAULT_SHOULD_FACTORIZE_LINEARS,
+    )
+    min_channels: int = hp.optional(
+        doc='Minimum number of channels in a Conv2d module'
+        ' for it to be factorized.',
+        default=_DEFAULT_MIN_CHANNELS,
+    )
+    latent_channels: float = hp.optional(
+        doc='Number of channels in factorized convolution latent representations',
+        default=_DEFAULT_LATENT_CHANNELS,
+    )
+    min_features: int = hp.optional(
+        doc='Minimum number of features in a Linear module'
+        ' for it to be factorized.',
+        default=_DEFAULT_MIN_FEATURES,
+    )
+    latent_features: float = hp.optional(
+        doc='Number of features in factorized linear latent representations',
+        default=_DEFAULT_LATENT_FEATURES,
+    )
+
+    def initialize_object(self) -> 'Factorize':
+        return Factorize(**asdict(self))
+
+
+class Factorize(Algorithm):
+
+    def __init__(self,
```

Comment:
Done. Also, in case it's relevant in the future, I checked and sphinx is smart here and puts the values of the constants in the docs instead of their names